dm/master/simulator.go

```diff
@@ -0,0 +1,561 @@
+	if err != nil {
+		return "", "", errors.Annotatef(err, "route of table %s failed", dbutil.TableName(schema, table))
+	}
+	return matchRoute, dbutil.TableName(targetSchema, targetTable), nil
+}
+
+// getSingleTableRouteResult gets matched route name and target table from one MySQL instance
+// returns filtered, matchRoute, matchTable, error
+func getSingleTableRouteResult(caseSensitive bool, tableQuery string, bwRules *filter.Rules, routeMap map[string]*router.TableRule) (bool, string, string, error) {
+	schema, table, err := utils.ExtractTable(tableQuery)
+	if err != nil {
+		return false, "", "", errors.Trace(err)
+	}
+	filtered, err := checkSingleBWRules(schema, table, caseSensitive, bwRules)
+	if err != nil {
+		return false, "", "", errors.Trace(err)
+	}
+	if filtered {
+		return true, "", "", nil
+	}
+	matchRoute, matchTable, err := getRouteName(caseSensitive, schema, table, routeMap)
+	if err != nil {
+		return false, "", "", errors.Trace(err)
+	}
+
+	return false, matchRoute, matchTable, nil
+}
+
+// getRoutePath gets table routes from one MySQL instance
+func getRoutePath(stCfg *config.SubTaskConfig) (*pb.SimulationResult, error) {
+	bwFilter, err := filter.New(stCfg.CaseSensitive, stCfg.BWList)
+	if err != nil {
+		return nil, errors.Annotatef(err, "build of black white filter for source %s failed", stCfg.SourceID)
+	}
+
+	r, err := router.NewTableRouter(stCfg.CaseSensitive, stCfg.RouteRules)
+	if err != nil {
+		return nil, errors.Annotatef(err, "build of router for source %s failed", stCfg.SourceID)
+	}
+
+	sourceDB, err := conn.DefaultDBProvider.Apply(stCfg.From)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer sourceDB.Close()
+
+	fetchedRouteTableMap, err := utils.FetchTargetDoTables(sourceDB.DB, bwFilter, r)
+	if err != nil {
+		return nil, errors.Annotatef(err, "routing from source %s failed", stCfg.SourceID)
+	}
+
+	routeTableMap := make(map[string]*pb.TableList, len(fetchedRouteTableMap))
+	// transfer filter.table to string
+	for targetName, routeTableList := range fetchedRouteTableMap {
+		sourceTableStringList := make([]string, 0, len(routeTableList))
+		for _, sourceTable := range routeTableList {
+			sourceTableStringList = append(sourceTableStringList, sourceTable.String())
+		}
+		routeTableMap[targetName] = &pb.TableList{Tables: sourceTableStringList}
+	}
+
+	return &pb.SimulationResult{
+		SourceID:      stCfg.SourceID,
+		SourceAddr:    getSourceAddr(stCfg.From),
+		RouteTableMap: routeTableMap,
+	}, nil
+}
+
+// getDoIgnoreTables gets do tables and ignore tables from one MySQL instance
+func getDoIgnoreTables(stCfg *config.SubTaskConfig) (*pb.SimulationResult, error) {
+	bwFilter, err := filter.New(stCfg.CaseSensitive, stCfg.BWList)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	sourceDB, err := conn.DefaultDBProvider.Apply(stCfg.From)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer sourceDB.Close()
+
+	mapping, err := utils.FetchAllDoTables(sourceDB.DB, nil)
```

Comment:
Is it valid to pass a `nil` filter?

Reply:
I think it's valid. It means that this `sourceDB` has no schema and table.